pyhf/pdf.py

```diff
@@ -131,6 +137,173 @@ def __init__(self, spec, **config_kwargs):
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'shapefactor':
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'shapesys':
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'staterror':
+                    modspec.setdefault('data',{})['uncrt'] = []
+                    modspec.setdefault('data',{})['mask']  = []
+                mega_mods.setdefault(s,{})[m] = copy.deepcopy(modspec)
+                
+        mega_samples = {}
+        for s in self.do_samples:
+            mega_nom = []
+            for c in self.do_channels:
+                defined_samp = helper.get(c,{}).get(s)
+                defined_samp = None if not defined_samp else defined_samp[1]
+                nom = defined_samp['data'] if defined_samp else [0.0]*self.channel_nbins[c]
+                mega_nom += nom
+                defined_mods = {x['name']:x for x in defined_samp['modifiers']} if defined_samp else {}
+                for m,mtype in self.do_mods:
+                    thismod = defined_mods.get(m)
+                    if mtype == 'histosys':
+                        lo_data = thismod['data']['lo_data'] if thismod else nom
+                        hi_data = thismod['data']['hi_data'] if thismod else nom
+                        maskval = True if thismod else False
+                        mega_mods[s][m]['data']['lo_data'] += lo_data
+                        mega_mods[s][m]['data']['hi_data'] += hi_data
+                        mega_mods[s][m]['data']['mask']    += [maskval]*len(nom) #broadcasting
+                        pass
+                    elif mtype == 'normsys':
+                        maskval = True if thismod else False
+                        lo_factor = thismod['data']['lo'] if thismod else 1.0
+                        hi_factor = thismod['data']['hi'] if thismod else 1.0
+                        mega_mods[s][m]['data']['lo']   += [lo_factor]*len(nom) #broadcasting
+                        mega_mods[s][m]['data']['hi']   += [hi_factor]*len(nom)
+                        mega_mods[s][m]['data']['mask'] += [maskval]  *len(nom) #broadcasting
+                    elif mtype == 'normfactor':
+                        maskval = True if thismod else False
+                        mega_mods[s][m]['data']['mask'] += [maskval]*len(nom) #broadcasting
+                    elif mtype == 'staterror':
+                        uncrt = thismod['data'] if thismod else [0.0]*len(nom)
+                        maskval = [True if thismod else False]*len(nom)
+                        mega_mods[s][m]['data']['mask']  += maskval
+                        mega_mods[s][m]['data']['uncrt'] += uncrt
+                    elif mtype == 'shapefactor':
+                        maskval = True if thismod else False
+                        mega_mods[s][m]['data']['mask'] += [maskval]*len(nom) #broadcasting
+                    else:
```

Comment:
for shapefactors and staterrors we just keep a mask as to which bins are affected